tmt/steps/__init__.py

```diff
@@ -392,12 +404,145 @@ def go(self):
+        except KeyError:
+            return []
+
+    @classmethod
+    def _parse_phases(cls, step):
+        """ Parse options and store phase order """
+        phases = dict()
+        options = cls._opt('step')
+
+        # Use the end of the last enabled step if no --step given
+        if not options:
+            last_enabled_step = list(step.plan.steps())[-1]
+            phases[last_enabled_step.name] = [PHASE_END]
+            return phases
+
+        # Process provided options
+        for option in options:
+            # Parse the step:phase format
+            matched = re.match(r'(\w+)(:(\w+))?', option)
+            if matched:
+                step_name, _, phase = matched.groups()
+            if not matched or step_name not in STEPS:
+                raise tmt.utils.GeneralError(f"Invalid step '{option}'.")
+            # Check phase format, convert into int, use end by default
+            try:
+                phase = int(phase)
+            except TypeError:
+                phase = PHASE_END
+            except ValueError:
+                # Convert 'start' and 'end' aliases
+                try:
+                    phase = dict(start=PHASE_START, end=PHASE_END)[phase]
+                except KeyError:
+                    raise tmt.utils.GeneralError(f"Invalid phase '{phase}'.")
+            # Store the phase for given step
+            try:
+                phases[step_name].append(phase)
+            except KeyError:
+                phases[step_name] = [phase]
+        return phases
+
+
+class Reboot(ActionStep):
+    """ Reboot guest """
+
+    # Which kind of reboot
+    _hard = False
+
+    # True if interactive login enabled
+    _enabled = False
+
+    # guest name provided via CLI
+    _guest_name = None
+
+    def __init__(self, step, order):
+        """ Initialize relations, store the reboot order """
+        super().__init__(parent=step, name='reboot')
+        self.order = order
+
+    @classmethod
+    def command(cls, method_class=None, usage=None):
+        """ Create the reboot command """
+        @click.command()
+        @click.pass_context
+        @click.option(
+            '--hard', metavar='REBOOT_HARD', is_flag=True,
+            help='Hard reboot of provisioned machine.')
+        @click.option(
+            '--name', metavar='MACHINE_NAME', multiple=False,
+            default=None, help='Provisioned machine name.')
+        def reboot(context, **kwargs):
+            """
+            Reboot machine.
+            """
+            Reboot._save_context(context)
+            Reboot._enabled = True
+
+        return reboot
+
+    @classmethod
+    def plugins(cls, step):
+        """ Return list of reboot instances for given step """
+        if not Reboot._enabled:
+            return []
+        return [Reboot(step, phase) for phase in cls.phases(step)]
+
+    def go(self, *args, **kwargs):
+        """ Reboot the guest(s) """
+
+        self._guest_name = name
```

Comment:
I may be blind but where does the `name` variable come from? Isn't it undefined? Also if it is only used in this method, I don't think it's necessary to set is a an instance attribute. I think you want to use `self.opt('name')` here to get the name from the saved context.